### PR TITLE
introducing native php serialize() support for Messenger transport

### PR DESCRIPTION
**Pull request by weaverryan**
| Q             | A
| ------------- | ---
| Branch?       | master
| Bug fix?      | yes and no
| New feature?  | yes and no
| BC breaks?    | maybe (yes if we change the default)
| Deprecations? | no
| Tests pass?   | yes
| Fixed tickets | #29163
| License       | MIT
| Doc PR        | TODO!

Messenger currently uses the Serialize to serialize to JSON and then unserialize. This creates a lot of issues:

1) The default serializer requires you to have getter & setter method (or public properties) for them to be serialized. This makes it easy for data to disappear. I've seen several really smart people have this problem.

2) Related to the above, the forced getters/setters (and no required constructor args) force you to design your message classes around this.

It's not that the serializer is doing a bad job - it's just not the right use-case for it.

This PR proposes simply using `serialize()` and `unserialize()`. This is the behavior we want: we want to put objects to sleep and wake them back up.

I believe the original reason we did not do this was so that we could export "generic JSON", in case we wanted other workers (not our Symfony app) to consume the messages. But, that's an edge case, and could still be accomplished by creating your own serializer.

Btw, Laravel uses `serialize()` as does Enqueue for (un)serializing Event objects. We're making our life more difficult for no benefit.

Cheers!

### CODE REVIEW

File: src/Symfony/Bundle/FrameworkBundle/Resources/config/messenger.xml [hunk 1]
```diff
@@ -24,6 +24,8 @@
         </service>
         <service id="Symfony\Component\Messenger\Transport\Serialization\SerializerInterface" alias="messenger.transport.serializer" />
 
+        <service id="messenger.transport.native_php_serializer"  class="Symfony\Component\Messenger\Transport\Serialization\Serializer" />
```
Comment (fancyweb) [inline src/Symfony/Bundle/FrameworkBundle/Resources/config/messenger.xml]: Extra space before the "class" + wrong class.

**Comment by fancyweb** [comment]
Also you need to modify https://github.com/symfony/symfony/blob/master/src/Symfony/Bundle/FrameworkBundle/Tests/DependencyInjection/ConfigurationTest.php#L298.

**Comment by xabbuh** [comment]
Given the issues we currently face in the Security component related to PHP's `serialize()` implementation (see for example #29951) I am not sure if we should actively support this.

**Review by nicolas-grekas — APPROVED** [review]
Definitely :+1:, this issue is really killing productivity right now for no good reason. Security is not an issue here: messages are not user input by themselves.

**Comment by nicolas-grekas** [comment]
#29951 is about the Serializable interface and parent::serialize(). We should activity discourage using it anywhere BTW.

File: src/Symfony/Component/Messenger/Transport/Serialization/PhpSerializer.php [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+ * file that was distributed with this source code.
+ */
+
+namespace Symfony\Component\Messenger\Transport\Serialization;
+
+use Symfony\Component\Messenger\Envelope;
+use Symfony\Component\Messenger\Exception\InvalidArgumentException;
+
+/**
+ * @author Ruyan Weaver<ryan@symfonycasts.com>
```
Comment (stloyd) [inline src/Symfony/Component/Messenger/Transport/Serialization/PhpSerializer.php]: Typo here 😄

**Comment by fabpot** [comment]
That basically reverts #28270, right?

**Comment by javiereguiluz** [comment]
I haven't used this component, but as an outsider, compatibility with third-party systems look like a huge and important feature to me. Are you sure it's OK to make us incompatible with anything not related to PHP?

Somewhat related: for the future, we could consider using Google's Protobuf, which is what most cool and modern apps use nowadays (https://developers.google.com/protocol-buffers/docs/reference/php-generated).

**Comment by dkarlovi** [comment]
Both your points are about the serializer doing a bad job with DTOs. Why not improve the Serializer to support DTOs better instead, it fixes this problem by proxy, but also improves the other component.

**Comment by dunglas** [comment]
In my opinion we should support both, and indeed default to PHP's `serialize()`. @weaverryan's arguments make sense, but only when the consumer is written in PHP. It often makes sense to use other languages for async consumers (Go or C for computations, Python for Machine Learning...), in such cases consuming JSON is easier than parsing the PHP internal format.

Protobuf is also very interesting for this kind of use cases (but is not as popular than JSON).

**Comment by dunglas** [comment]
@dkarlovi we need to improve DTO support indeed (even if it has already been dramatically improved in 4.1/4.2), but it's not always the right tool anyway for messenger: `serialize` preserves private and protected states. It makes no senses to add this kind of features to the Symfony serializer (its goal is to transform the object in a **public** representation).

**Comment by alcaeus** [comment]
I agree that using the Serializer Component makes has some drawbacks compared to using `serialize`.  On the other hand, it ensures that developers can choose a different serialisation method depending on their environment.

I also agree with @dunglas that the only useful way to fix this is to make it configurable: Symfony has always encouraged a pluggable system where you could easily modify the system to your specific needs while providing a sensible default setup. I believe that's what should be done here: allow using any serialiser but default to one using PHPs built-in serialisation functionality.

**Comment by dkarlovi** [comment]
@dunglas don't really understand: the message in Messenger context **is** in its public context.

**Comment by dunglas** [comment]
@dkarlovi not really:

* private states are preserved when using synchronous handlers
* they aren't when configure a transport

It hurts DX (especially for those used to how Laravel works). Messaging could be considered internal (as done when using a sync handler) from an app level. The serialization is just an implementation detail, and so preserving states would improve both DX and performance (it just works).

I would say that it's different if the message is intended to the current app (like when using a messenger handler) => private, or for another app (like one written in another language) => public.

**Comment by nicolas-grekas** [comment]
> especially for those used to how Laravel works

no need to write this. I've no idea what Laravel is, yet I spent an afternoon debugging that transport thing until I created #29163. We have a serious issue here.

The serialization mechanism is already "pluggable", here we're talking about changing the default.
I don't remember if we have a DI config integration to leverage this pluggability?

**Comment by ostrolucky** [comment]
Key word is "default serializer". Symfony serializer or symfony property accessor requiring getters/setters or public props is not a problem for messenger only. See eg https://github.com/nelmio/alice/blob/b7a0e6652570123b349ffa78dd7392cb58882013/src/PropertyAccess/ReflectionPropertyAccessor.php. You are not solving this problem by dropping it. Neither doctrine hydrator, automapper or nelmio fixtures have this limitation and they don't use native serialization. Seems it's time to revisit this convention of serializer and think about adding possibility to serialize such props as well. Cc @dunglas

**Comment by dunglas** [comment]
@nicolas-grekas I mean when you use Enqueue (it's the use case described by @weaverryan in the PR description), messaging is considered something purely internal and transparent. You don't even have to know that the message is serialized. It's "magic", but it works. When switching to Messenger, you have to think about using public properties, because JSON.

> Seems it's time to revisit this convention of serializer and think about adding possibility to serialize such props as well

It's easy to do (as showcased by the code you linked).

> they don't use native serialization

But why? native serialization is faster, and easy to use.

**Comment by dkarlovi** [comment]
> I would say that it's different if the message is intended to the current app (like when using a messenger handler) => private, or for another app (like one written in another language) => public.

It depends how you define "private", I guess. From Messenger's POV, the message passed is definitely in its **public representation** since it's going out of the local app's context onto a transport, it's irrelevant if the consumer is the same app again, that will be determined long **after** you've already serialized it.

**Comment by dunglas** [comment]
> I don't remember if we have a DI config integration to leverage this pluggability?

We have such kind of configs in other components at least. Could be something like:

```yaml
messenger:
    # ...
    serialize: private (default) # can also be public (JSON) or a service name
```

**Comment by Pierstoval** [comment]
> We have such kind of configs in other components at least. Could be something like:
> 
> ```yaml
> messenger:
>     # ...
>     serialize: private (default) # can also be public (JSON) or a service name
> ```

I agree with this point and instead of private/public I'd suggest `native|{service_id}`, referring to services sounds like the best IMO

**Comment by nicolas-grekas** [comment]
> I don't remember if we have a DI config integration to leverage this pluggability?

the answer is yes, I just needed to look at the first file in the PR :)
There is nothing else to do here, by default we consider messages "private", which we should.
And ppl who need interop with third-party apps have more work to do.
That's how things should be IMHO.

**Comment by Pierstoval** [comment]
> And ppl who need interop with third-party apps have more work to do.

This will probably continue to be debated, we might need "stats" about what proportion of apps use or don't use 3rd-party apps 😕

**Comment by nicolas-grekas** [comment]
Let's agree the DX is seriously broken, up to the point where advocating the component for the masses is questionable currently.

This PR is a serious solution to the problem.
The Serializer component is really nice when compat with other techno is needed.
But that comes AFTER.

**Comment by dkarlovi** [comment]
> And ppl who need interop with third-party apps have more work to do.

I'm fine with this conclusion, if you need to cover this use case, you shouldn't be too wary of it not being covered by default, as long as you can actually do it with condiguration.

**Comment by ciaranmcnulty** [comment]
We're currently looking at using messenger for this use case of emitting messages to a third party but I'd always assumed we'd be writing a serialiser explicitly for that case (for instance having the FQCN in the headers of the AMQP message doesn't make sense)

It'd be nice for 'soft BC' to have the JSON serlialiser as a config option even if the default is switched to native and users have to add something to their config to maintain the JSON behaviour.

**Comment by ostrolucky** [comment]
I'm ok with this solution since it's just about default which can be (hopefully) easily switched. But this is not a correct solution to reasons given for this change. Correct solution is to improve Serializer. Better justification for this change is performance and simplification.

**Comment by dkarlovi** [comment]
Just to reiterate my points made in a Slack thread: treating messages as "private" (or "internal") is misleading since they're leaving the current context (when sending) and coming from a source which shouldn't just be trusted by default.

**The transport is an external resource** and any messages going to it / coming from it should be treated as such, security and all. The question is, can a bad actor somehow trick the consuming Symfony app into doing something it didn't expect to do by placing a message with a carefully constructed payload (ie. all the security implications of `unserialize` apply here).

**Comment by dunglas** [comment]
Here is the current config:

```yaml
    ```messenger:
        enabled:              false
        routing:

            # Prototype
            message_class:
                senders:              []
                send_and_handle:      false
        serializer:
            id:                   messenger.transport.symfony_serializer
            format:               json
            context:

                # Prototype
                name:                 ~
        transports:

            # Prototype
            name:
                dsn:                  ~
                options:              []
        default_bus:          null
        buses:

            # Prototype
            name:
                default_middleware:   true # One of true; false; "allow_no_handlers"
                middleware:

                    # Prototype
                    -
                        id:                   ~ # Required
                        arguments:            []```
```

I suggest to move the `serializer` config under the `message_class` key. Then it will be possible to configure a specific serializer (internal, public, porotbuf...) per message class.

**Comment by nicolas-grekas** [comment]
> I suggest to move the serializer config under the routing.message_class key.

I agree it's strange to have it directly under messenger. Under "transport" might be a better fit.
But let's move this elsewhere and keep the focus of this PR :)

**Comment by weaverryan** [comment]
Summarizing a few important points:

1) The serialization is configurable and will always be configurable.

2) There is no added security concern as the messages are currently serialized to JSON, but they are unserialized back to objects by using some class name metadata stored in the headers of the message. If a bad actor is able to add messages to your queue, you have a problem either way.

Better DX for the 99% use-case, no added security problems, easy configurability for the 1% :).

**Review by chalasr — APPROVED** [review]

**Review by dunglas — APPROVED** [review]

File: src/Symfony/Bundle/FrameworkBundle/Resources/config/messenger.xml [hunk 3]
```diff
@@ -24,6 +24,8 @@
         </service>
         <service id="Symfony\Component\Messenger\Transport\Serialization\SerializerInterface" alias="messenger.transport.serializer" />
 
+        <service id="messenger.transport.native_php_serializer" class="Symfony\Component\Messenger\Transport\Serialization\Serializer" />
```
Comment (xabbuh) [inline src/Symfony/Bundle/FrameworkBundle/Resources/config/messenger.xml]:
```suggestion
        <service id="messenger.transport.native_php_serializer" class="Symfony\Component\Messenger\Transport\Serialization\PhpSerializer" />
```

**Review by xabbuh — CHANGES_REQUESTED** [review]

**Review by xabbuh — APPROVED** [review]

**Review by nicolas-grekas — CHANGES_REQUESTED** [review]
Oh, one last thing:
`Symfony\Component\Messenger\Transport\Serialization\Serializer::create()` should return that new serializer, isn't it?  Maybe not actually, can anyone confirm?

**Comment by nicolas-grekas** [comment]
Maybe not actually, can anyone confirm?

**Comment by weaverryan** [comment]
> Oh, one last thing:
Symfony\Component\Messenger\Transport\Serialization\Serializer::create() should return that new serializer, isn't it? Maybe not actually, can anyone confirm?

I think no, because that's a factory to literally create an instance of THAT class. So, it should continue to create an instance of itself. However, there were 4 places in the code that used this factory function to create a default serializer if none was passed. I change those to create the new `PhpSerializer`. I think we should consistently use the same default serializer everywhere.

File: src/Symfony/Component/Messenger/Transport/Serialization/PhpSerializer.php [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+
+namespace Symfony\Component\Messenger\Transport\Serialization;
+
+use Symfony\Component\Messenger\Envelope;
+use Symfony\Component\Messenger\Exception\InvalidArgumentException;
+
+/**
+ * @author Ryan Weaver<ryan@symfonycasts.com>
+ *
+ * @experimental in 4.2
```
Comment (jdreesen) [inline src/Symfony/Component/Messenger/Transport/Serialization/PhpSerializer.php]: Shouldn't this be `4.3`?

Comment (weaverryan) [reply src/Symfony/Component/Messenger/Transport/Serialization/PhpSerializer.php]: Hmm, actually - not sure. The component is experimental in 4.2 - it won’t be (as of now in 4.3). I added this for consistency with the rest of the component. Either all will be updates at the same time to 4.3 or (hopefully) all will be removed.

**Comment by akorz** [comment]
Hello,

I have framework.yaml
```
    messenger:
        serializer:
            id: 'native_php_serializer'
```

but get error

```
In ArrayNode.php line 331:
                                                                                                                                      
  Unrecognized option "id" under "framework.messenger.serializer". Available options are "default_serializer", "symfony_serializer".  
```

What should I fix in my config? Thanks

**Comment by xabbuh** [comment]
This should work (assuming that `native_php_serializer` is the id of a service that you defined):

```yaml
framework:
    messenger:
        serializer:
            default_serializer: 'native_php_serializer'
```

**Comment by akorz** [comment]
@xabbuh I was thinking it should be defined somewhere in component? 

Now I have something else

```
Symfony\Component\DependencyInjection\Exception\ServiceNotFoundException: The service "enqueue.messenger_transport.factory" has a dependency on a non-existent service "messenger.transport.serializer".
```

I could not find any docs about it :(

**Comment by xabbuh** [comment]
I guess you can best use one of the [support channels](https://symfony.com/support) to get help setting things up properly.

**Comment by mtomala** [comment]
@weaverryan 
There is a bug (probably) in blog post about this feature:
https://symfony.com/blog/new-in-symfony-4-3-native-php-serialization-for-messenger

It should be `default_serializer:` instead of `id:`

**Comment by javiereguiluz** [comment]
@mtomala when you and @weaverryan decide if this is a bug or not ... please ping me so I can update the blog post. Thanks.

**Comment by kunicmarko20** [comment]
At the time when this feature was added https://github.com/symfony/symfony/pull/29958 it was still `id` now it is `default_serializer`

**Comment by ogizanagi** [comment]
@javiereguiluz: Since https://github.com/symfony/symfony/pull/30628, the serializer to use is actually configurable per transport.

The `messenger.serializer` section remains to configure both native (default) and symfony serializers:

https://github.com/symfony/symfony/blob/dca9325e61cbe616ec2d5c83e83af0e948eba3a5/src/Symfony/Bundle/FrameworkBundle/DependencyInjection/Configuration.php#L1155-L1176

but each transport can specify the `messenger.transport.[name].serializer` option to specify the service id to use.
https://github.com/symfony/symfony/blob/dca9325e61cbe616ec2d5c83e83af0e948eba3a5/src/Symfony/Bundle/FrameworkBundle/DependencyInjection/Configuration.php#L1189

**Pull request merged**